scheduler/src/cook/kubernetes/api.clj

```diff
@@ -0,0 +1,316 @@
+                                              .getName))
+                                        (.getItems current-pods))]
+    (log/info "Updating current-pods-atom with pods" (keys pod-name->pod))
+    (log/error "TODO: We should have two atoms, one for cook pods (that feeds into the controller) and one for
+    all pods (that feeds into machine utilization then into offer generation)")
+    (reset! current-pods-atom pod-name->pod)
+    (let [watch (WatchHelper/createPodWatch api-client (-> current-pods
+                                                           .getMetadata
+                                                           .getResourceVersion))]
+      (.submit kubernetes-executor ^Callable
+      (fn []
+        (try
+          (handle-watch-updates current-pods-atom watch (fn [p] (-> p .getMetadata .getName)) pod-callback)
+          (catch Exception e
+            (log/error "Error during watch: " (-> e .toString))
+            (.close watch)
+            (initialize-pod-watch api-client current-pods-atom pod-callback))))))))
+
+(defn initialize-node-watch [^ApiClient api-client current-nodes-atom]
+  (let [api (CoreV1Api. api-client)
+        current-nodes (.listNode api
+                                 nil ; includeUninitialized
+                                 nil ; pretty
+                                 nil ; continue
+                                 nil ; fieldSelector
+                                 nil ; labelSelector
+                                 nil ; limit
+                                 nil ; resourceVersion
+                                 nil ; timeoutSeconds
+                                 nil ; watch
+                                 )
+        node-name->node (pc/map-from-vals (fn [^V1Node node]
+                                            (-> node .getMetadata .getName))
+                                          (.getItems current-nodes))]
+    (reset! current-nodes-atom node-name->node)
+    (let [watch (WatchHelper/createNodeWatch api-client (-> current-nodes .getMetadata .getResourceVersion))]
+      (.submit kubernetes-executor ^Callable
+      (fn []
+        (try
+          (handle-watch-updates current-nodes-atom watch (fn [n] (-> n .getMetadata .getName)) nil)
+          (catch Exception e
+            (log/warn e "Error during node watch")
+            (initialize-node-watch api-client current-nodes-atom))
+          (finally
+            (.close watch))))))))
+
+(defn to-double
+  [^Quantity q]
+  (-> q .getNumber .doubleValue))
+
+(defn convert-resource-map
+  [m]
+  {:mem (if (get m "memory")
+          (-> m (get "memory") to-double (/ (* 1024 1024)))
+          0.0)
+   :cpus (if (get m "cpu")
+           (-> m (get "cpu") to-double)
+           0.0)})
+
+(defn get-capacity
+  [node-name->node]
+  (pc/map-vals (fn [^V1Node node]
+                 (-> node .getStatus .getCapacity convert-resource-map))
+               node-name->node))
+
+(defn get-consumption
```

Comment:
clojuredoc. Also might be worth giving the shape of the input&output dictionary. Also probably a punchlist item to figure out if requests and limit behave as we expect for accounting as far as launching.

Reply:
👍 .  I don't think we need a followup task, it seems pretty clear: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container

Reply:
Partially done: @pschorf  can you polish the comment?

Reply:
Done